test_util/provide_aws.py

```diff
@@ -0,0 +1,152 @@
+        self.region = region
+        self.session = boto3.session.Session(
+                aws_access_key_id=aws_access_key_id, aws_secret_access_key=aws_secret_access_key)
+
+    def client(self, name):
+        return self.session.client(service_name=name, region_name=self.region)
+
+    def resource(self, name):
+        return self.session.resource(service_name=name, region_name=self.region)
+
+    def create_key_pair(self, key_name):
+        """Retruns private key of newly generated pair
+        """
+        key = self.client('ec2').create_key_pair(KeyName=key_name)
+        return key['KeyMaterial']
+
+    def delete_key_pair(self, key_name):
+        self.client('ec2').delete_key_pair(KeyName=key_name)
+
+    def create_stack(self, name, template_url, user_parameters):
+        """Returns boto stack object
+        """
+        logging.info('Requesting AWS CloudFormation...')
+        cf_parameters = []
+        for k, v in user_parameters.items():
+            cf_parameters.append({'ParameterKey': k, 'ParameterValue': v})
+        return self.resource('cloudformation').create_stack(
+                StackName=name,
+                TemplateURL=template_url,
+                DisableRollback=True,
+                TimeoutInMinutes=30,
+                Capabilities=['CAPABILITY_IAM'],
+                Parameters=cf_parameters)
+
+
+class CFStack():
+    def __init__(self, region, aws_access_key_id, aws_secret_access_key):
+        self.boto_wrapper = BotoWrapper(region, aws_access_key_id, aws_secret_access_key)
+
+    def attach(self, name):
+        self.stack = self.boto_wrapper.resource('cloudformation').Stack(name)
+
+    def wait_for_stack_creation(self):
+        self.wait_for_transition('CREATE_IN_PROGRESS', 'CREATE_COMPLETE')
+
+    def wait_for_stack_deletion(self):
+        self.wait_for_transition('DELETE_IN_PROGRESS', 'DELETE_COMPLETE')
+
+    def wait_for_transition(self, start_status, end_status):
+        final_status, status_reason = self.wait_for_status_change(start_status)
+        assert end_status == final_status, "Expected status {} but found {}. Reason: {}".format(
+                end_status, final_status, status_reason)
+
+    @retrying.retry(wait_fixed=5000, stop_max_delay=AWS_WAITING_TIMEOUT*60*1000,
+                    retry_on_result=lambda res: res is False,
+                    retry_on_exception=lambda ex: False)
+    def wait_for_status_change(self, status):
+        """NOTE: need to get stack_id before looping on wait or else method will
+        raise an excetion if the stack deletes and it cannot find stack ID
+        """
+        logging.info('Waiting for stack to finish: '+status)
+        stack_id = self.stack.stack_id
+        current_status, current_reason = self.get_stack_status(stack_id)
+        if(current_status == status):
+            return False
+        return current_status, current_reason
+
+    def get_stack_status(self, stack_id=None):
```

Comment:
shouldn't take a stack_id, should always just use the class member (self.stack.stack_id).